### PR TITLE
Compile and fix_shebang in %python_install macros

### PR DESCRIPTION
**Pull request by danigm**

### CODE REVIEW

**Comment by mcepl** [comment]
That looks reasonable. What happens with testing build in your home branch?

**Comment by danigm** [comment]
> That looks reasonable. What happens with testing build in your home branch?

I've tested to build different python packages with this version of macros and it worked. I've tested python-pycurl and other packages that uses the %python_install macro. But I'm not sure if this change can produce any other side effect.

**Comment by mcepl** [comment]
Well, I don’t see anything, but probably our best hope is just to add more packages to your home project.

**Comment by bnavigator** [comment]
There are only a few packages using arguments in Factory and luckily none of them uses the backslash pattern found for configure and cmake calls:

```
%python_install \
  --myarg1
  --myarg2
```

In repo cloned from  https://github.com/bmwiedemann/openSUSE

```
openSUSE $ git grep 'python_install .'
packages/libk/libkdumpfile/libkdumpfile.spec:%define python_install python setup.py install --skip-build --root %{?buildroot}
packages/m/meson/meson.changes:  + Use python_build/python_install instead of
packages/p/python-XlsxWriter/python-XlsxWriter.spec:%{python_expand %$python_install && \
packages/p/python-aiofiles/python-aiofiles.changes:  * Replace %python_install with %pyproject_install
packages/p/python-configshell-fb/python-configshell-fb.spec:%python_install --skip-build
packages/p/python-mysql-connector-python/python-mysql-connector-python.spec:# when using python_install so use custom python_exec instead
packages/p/python-pycurl/python-pycurl.spec:%python_install --with-openssl
packages/p/python-pyo/python-pyo.spec:%python_install --use-jack --use-double
packages/p/python-pysvn/python-pysvn.spec:%{python_expand # python_build and python_install not applicable: not the standard setuptools workflow
packages/p/python-ruamel.yaml.clib/python-ruamel.yaml.clib.spec:%python_install --single-version-externally-managed
packages/p/python-ruamel.yaml/python-ruamel.yaml.spec:%python_install --single-version-externally-managed
packages/p/python-spidev/python-spidev.spec:%python_install -O1 --skip-build
```

**Comment by bnavigator** [comment]
Probably need to fix python-XlsXWriter then

**Comment by danigm** [comment]
Following the usage in python-XlsxWriter I've detected other packages that will be broken, I'm trying to collect different user cases in this test project: https://build.opensuse.org/project/show/home:dgarcia:fix-shebang

**Comment by bnavigator** [comment]
Ah you are right, there is more, including rpmconf with the backslash pattern:

```
$ git grep 'python.*_install .'
packages/b/benji/benji.spec:%python3_install --single-version-externally-managed
packages/g/gnuhealth-client/gnuhealth-client.spec:%python3_install --prefix=%{_prefix} --root=%{buildroot}
packages/g/gnuhealth/gnuhealth.spec:  %python3_install --prefix=%_prefix --root=%buildroot
packages/libk/libkdumpfile/libkdumpfile.spec:%define python_install python setup.py install --skip-build --root %{?buildroot}
packages/libk/libkdumpfile/libkdumpfile.spec:%define python3_install python3 setup.py install --skip-build --root %{?buildroot}
packages/m/meson/meson.changes:  + Use python_build/python_install instead of
packages/o/oci-cli/oci-cli.changes:- Replace `python3 setup.py install` with %python3_install in %install section
packages/p/proteus/proteus.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/p/python-XlsxWriter/python-XlsxWriter.spec:%{python_expand %$python_install && \
packages/p/python-aiofiles/python-aiofiles.changes:  * Replace %python_install with %pyproject_install
packages/p/python-compizconfig/python-compizconfig.spec:%python_expand %make_install -C "build-$python"
packages/p/python-configshell-fb/python-configshell-fb.spec:%python_install --skip-build
packages/p/python-linux-procfs/python-linux-procfs.changes:- Use python3_install and python3_build, this package is only
packages/p/python-mysql-connector-python/python-mysql-connector-python.spec:# when using python_install so use custom python_exec instead
packages/p/python-pycurl/python-pycurl.spec:%python_install --with-openssl
packages/p/python-pyo/python-pyo.spec:%python_install --use-jack --use-double
packages/p/python-pysvn/python-pysvn.spec:%{python_expand # python_build and python_install not applicable: not the standard setuptools workflow
packages/p/python-ruamel.yaml.clib/python-ruamel.yaml.clib.spec:%python_install --single-version-externally-managed
packages/p/python-ruamel.yaml/python-ruamel.yaml.spec:%python_install --single-version-externally-managed
packages/p/python-scipy/python-scipy.spec:%{$python_pyproject_install --prefix %{p_prefix}}
packages/p/python-spidev/python-spidev.spec:%python_install -O1 --skip-build
packages/r/rpmconf/rpmconf.spec:%python3_install \
packages/t/tensorflow-lite/tensorflow-lite.spec:%python3_install --install-lib=%{python3_sitearch}
packages/t/trytond/trytond.spec:%python3_install --prefix=%{_prefix} --root=%{buildroot}
packages/t/trytond_account/trytond_account.spec:%python3_install --prefix=%{_prefix} --root=%{buildroot}
packages/t/trytond_account_invoice/trytond_account_invoice.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_account_invoice_stock/trytond_account_invoice_stock.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_account_product/trytond_account_product.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_company/trytond_company.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_country/trytond_country.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_currency/trytond_currency.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_party/trytond_party.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_product/trytond_product.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_purchase/trytond_purchase.spec:%python3_install --prefix=%_prefix --root=%buildroot 
packages/t/trytond_purchase_request/trytond_purchase_request.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_stock/trytond_stock.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_stock_lot/trytond_stock_lot.spec:%python3_install --prefix=%_prefix --root=%buildroot
packages/t/trytond_stock_supply/trytond_stock_supply.spec:%python3_install --prefix=%_prefix --root=%buildroot
```

**Comment by bnavigator** [comment]
Comment from @DimStar77 in https://build.opensuse.org/request/show/1116747:

> This breaks fun assumptions on rpmlint-mini - which is certainly a hack, but works
>
> i.e rpmlint-mini is still based on python3.10 (don't ask me why... maintainers choice) - but it copies that actually into /opt/testing/bin as python3; this new macro to fix shebang goes and changes the python3 shebang to a python3.11 - even though we have python 3.10 in opt/testing/bin

**Comment by bnavigator** [comment]
... which was "fixed" by https://build.opensuse.org/request/show/1116869.

But I think it shows that there are way too many assumptions in place which are not well-enough documented.

**Comment by danigm** [comment]
The `%python3_install --prefix=%_prefix --root=%buildroot` that's used in a lot of packages is a redundancy, the current macro has that prefix/root option. I'm not sure what's the effect with the current macro to have these options duplicated, if it's applied the first one or the latest one.

**Comment by mcepl** [comment]
> The `%python3_install --prefix=%_prefix --root=%buildroot` that's used in a lot of packages is a redundancy, the current macro has that prefix/root option. I'm not sure what's the effect with the current macro to have these options duplicated, if it's applied the first one or the latest one.

It should be the second one (so it could be modified), but if it is, I have no idea.

**Comment by danigm** [comment]
I think this is ready to be merged. Maybe we will find more packages that are broken when this lands on tumbleweed, but I think this is something needed and any issue with the macro should be fixed in the package.

File: flavor.in [hunk 1]
```diff
@@ -100,11 +103,13 @@ for f in sys.argv[1:]: \
 done
 
 %#FLAVOR#_fix_shebang \
-for f in %{buildroot}%{_bindir}/*; do \
-  [ -f $f ] && sed -i "1s@#!.*python.*@#!$(realpath %__#FLAVOR#)@" $f \
-done; \
-for f in %{buildroot}%{_sbindir}/*; do \
-  [ -f $f ] && sed -i "1s@#!.*python.*@#!$(realpath %__#FLAVOR#)@" $f \
+%#FLAVOR#_fix_shebang_path %{buildroot}%{_bindir}/* \
+%#FLAVOR#_fix_shebang_path %{buildroot}%{_sbindir}/*
```
Comment (bnavigator) [inline flavor.in:107]:
```suggestion
%#FLAVOR#_fix_shebang_path %{buildroot}%{_bindir}/*  %{buildroot}%{_sbindir}/*
```

**Pull request merged**